ITensorGaussianMPS/src/gmps.jl

```diff
@@ -196,44 +485,90 @@ stored in a structure as a set of Givens rotations.
 
 The correlation matrix should be Hermitian, and will be treated as if it itensor
 in the algorithm.
+
+If `is_bcs`, the correlation matrix is assumed to be in interlaced format:
+Λ[2*i-1:2*i,2*j-1:2*j]=[[c_i c_j^dagger , c_i c_j ], [c_i^dagger c_j^dagger,c_i^dagger c_j]]
+Note that this may not be the standard choice in the literature, but it is internally
+consistent with the format of single-particle Hamiltonians and Slater determinants employed.
 """
+###Backward Compatibility
+
+#correlation_matrix_to_gmps(
+#  Λ0::AbstractMatrix; eigval_cutoff::Float64=1e-8, maxblocksize::Int=size(Λ0.data, 1)
+#)= correlation_matrix_to_gmps(
+#    Λ0; eigval_cutoff=eigval_cutoff, minblocksize=1, maxblocksize=maxblocksize
+#  )
+
 function correlation_matrix_to_gmps(
-  Λ0::AbstractMatrix{ElT}; eigval_cutoff::Float64=1e-8, maxblocksize::Int=size(Λ0, 1)
-) where {ElT<:Number}
-  Λ = Hermitian(Λ0)
-  N = size(Λ, 1)
+  Λ0::AbstractMatrix;
+  eigval_cutoff::Float64=1e-8,
+  minblocksize::Int=1,
+  maxblocksize::Int=size(Λ0.data, 1),
+)
+  return correlation_matrix_to_gmps(
+    ConservingNf(Λ0);
+    eigval_cutoff=eigval_cutoff,
+    minblocksize=minblocksize,
+    maxblocksize=maxblocksize,
+  )
+end
+
+function correlation_matrix_to_gmps(
+  Λ0::T;
+  eigval_cutoff::Float64=1e-8,
+  minblocksize::Int=1,
+  maxblocksize::Int=size(Λ0.data, 1),
+) where {T<:AbstractSymmetry}
+  Λ = T(Hermitian(copy(Λ0.data)))
+  ElT = eltype(Λ.data)
   V = Circuit{ElT}([])
-  ns = Vector{real(ElT)}(undef, N)
   err_tot = 0.0
-  for i in 1:N
-    blocksize = 0
-    n = 0.0
+  Λ = maybe_drop_pairing_correlations(Λ)
+  N = size(Λ.data, 1)
+  #calctype = typeof(Λ)
+  ns = set_data(Λ, Vector{real(ElT)}(undef, N))
+  for i in 1:div(N, site_stride(Λ))
     err = 0.0
-    p = Int[]
-    uB = 0.0
-    for blocksize in 1:maxblocksize
-      j = min(i + blocksize, N)
-      ΛB = @view Λ[i:j, i:j]
-      nB, uB = eigen(Hermitian(ΛB))
-      p = sortperm(nB; by=entropy)
-      n = nB[p[1]]
-      err = min(n, 1 - n)
-      err ≤ eigval_cutoff && break
+    v, nB, err = isolate_subblock_eig(
+      Λ,
+      i;
+      eigval_cutoff=eigval_cutoff,
+      minblocksize=minblocksize,
+      maxblocksize=maxblocksize,
+    )
+    set_populations!(ns, nB, v, i)
+    if stop_gmps_sweep(v)
+      break
     end
-    err_tot += err
-    ns[i] = n
-    v = @view uB[:, p[1]]
     g, _ = givens_rotations(v)
-    shift!(g, i - 1)
+    replace_indices!(j -> j + site_stride(Λ) * (i - 1), g)
+
     # In-place version of:
     # V = g * V
-    lmul!(g, V)
-    Λ = Hermitian(g * Λ * g')
+    LinearAlgebra.lmul!(g, V)
+    Λ = set_data(Λ, Hermitian(g * Matrix(Λ.data) * g'))
   end
-  return ns, V
+  ###return non-wrapped populations for backwards compatibility
+  return ns.data, V
 end
 
+##ToDo: This restricts the data of AbstractSymmetry to have same eltype, but could also promote eltypes
+(x::AbstractSymmetry * y::AbstractSymmetry) =
```

Comment:
What about:
```julia
function (x::AbstractSymmetry * y::AbstractSymmetry)
  if !has_same_symmetry(x, y)
    error("Can't multiply two symmetric objects with different symmetries.")
  end
  return set_data(x, x.data * y.data)
end

has_same_symmetry(::AbstractSymmetry, ::AbstractSymmetry) = false
has_same_symmetry(::ConservesNf, ::ConservesNf) = true
has_same_symmetry(::ConservesNfParity, ::ConservesNfParity) = true
```
instead? That should be more general and extensible to other types wrapped by `AbstractSymmetry`.

Reply:
Yes, that looks better.